src/index/diskann/diskann.cc

```diff
@@ -433,32 +441,33 @@ DiskANNIndexNode<T>::Deserialize(const BinarySet& binset, const Config& cfg) {
             LOG_KNOWHERE_INFO_ << "Caching " << num_nodes_to_cache << " sample nodes around medoid(s).";
             if (prep_conf.use_bfs_cache.value()) {
                 LOG_KNOWHERE_INFO_ << "Use bfs to generate cache list";
+                std::vector<uint32_t> node_list;
                 if (TryDiskANNCall([&]() { pq_flash_index_->cache_bfs_levels(num_nodes_to_cache, node_list); }) !=
                     Status::success) {
                     LOG_KNOWHERE_ERROR_ << "Failed to generate bfs cache for DiskANN.";
                     return Status::diskann_inner_error;
                 }
-            } else {
-                LOG_KNOWHERE_INFO_ << "Use sample_queries to generate cache list";
-                if (TryDiskANNCall([&]() {
-                        pq_flash_index_->generate_cache_list_from_sample_queries(warmup_query_file, 15, 6,
-                                                                                 num_nodes_to_cache, node_list);
-                    }) != Status::success) {
-                    LOG_KNOWHERE_ERROR_ << "Failed to generate cache from sample queries for DiskANN.";
-                    return Status::diskann_inner_error;
+                if (node_list.size() > 0) {
+                    if (TryDiskANNCall([&]() { pq_flash_index_->load_cache_list(node_list); }) != Status::success) {
+                        LOG_KNOWHERE_ERROR_ << "Failed to load cache for DiskANN.";
+                        return Status::diskann_inner_error;
+                    }
                 }
+            } else {
+                LOG_KNOWHERE_INFO_ << "Use sample_queries and input queries to generate cache asynchronously.";
+                pq_flash_index_->set_async_cache_flag(true);
+                search_pool_->push([&, sample_file = warmup_query_file, cache_num = num_nodes_to_cache]() {
+                    try {
+                        pq_flash_index_->async_generate_cache_list_from_sample_queries(sample_file, 15, 6, cache_num);
+                    } catch (const std::exception& e) {
+                        LOG_KNOWHERE_ERROR_ << "DiskANN Exception: " << e.what();
+                    }
```

Comment:
Semaphore signaled or not when exception thrown?

Reply:
updated